### PR TITLE
Allow skip codegen build

### LINKED ISSUE #357: Try to allow building java without having to build codegen
It may be possible to use the last released codegen binary.

### PR DESCRIPTION
**Pull request by zhangkun83**
Resolves #357

- Default behavior is not changed.
- People who don't change the codegen nor the proto files can set settings property ``grpc.skip.codegen=true`` so that they don't need to set up C++ compilation.
- Check in all generated files under ``src/generated``.

@ejona86 please review

### CODE REVIEW

**Comment by nicolasnoble** [comment]
Can one of the admins verify this patch?

**Comment by ejona86** [comment]
It seems like project properties are exposed as `settings` in settings.gradle: https://gradle.org/docs/current/dsl/org.gradle.api.initialization.Settings.html

**Comment by ejona86** [comment]
Can we only commit the grpc generated code? Is there a reason to commit the protobuf generated code?

**Comment by zhangkun83** [comment]
> It seems like project properties are exposed as settings in settings.gradle: https://gradle.org/docs/current/dsl/org.gradle.api.initialization.Settings.html

Nice! I have updated the PR.

> Can we only commit the grpc generated code? Is there a reason to commit the protobuf generated code?

Good point. I am going to update google/protobuf-gradle-plugin/pull/7 to allow specifying generated source dir  for individual sourceSets so that we can keep protobuf generated code under ``generated-code`` while putting grpc generated code under ``src/generated``.

**Comment by ejona86** [comment]
Ah, I see. That answered my question. Having different source sets would be a little weird; it's fine as it is. I'll review this this morning.

File: lib/netty [hunk 1]
```diff
@@ -1 +1 @@
-Subproject commit 4ae8bdc6ec8d37106f386a7d667a9cd279a8f5d8
+Subproject commit 1cce998bb06f42ff47925390872dcc5c487bbf59
```
Comment (ejona86) [inline lib/netty]: This looks like a stray change, and will break the build. You likely didn't do a "git submodule update" and then did a "git commit -a". It'll be nice when we don't have to mess with the submodules...

File: settings.gradle [hunk 2]
```diff
@@ -21,8 +20,15 @@ project(':grpc-protobuf').projectDir = "$rootDir/protobuf" as File
 project(':grpc-protobuf-nano').projectDir = "$rootDir/protobuf-nano" as File
 project(':grpc-netty').projectDir = "$rootDir/netty" as File
 project(':grpc-testing').projectDir = "$rootDir/testing" as File
-project(':grpc-compiler').projectDir = "$rootDir/compiler" as File
 project(':grpc-integration-testing').projectDir = "$rootDir/integration-testing" as File
 project(':grpc-all').projectDir = "$rootDir/all" as File
 project(':grpc-benchmarks').projectDir = "$rootDir/benchmarks" as File
 project(':grpc-examples').projectDir = "$rootDir/examples" as File
+
+if (settings.hasProperty('grpc.skip.codegen')
```
Comment (ejona86) [inline settings.gradle]: If this is for new contributors, do we want to have the default to skip grpc codegen? The very few people that actually touch codegen would then be the only ones who need to set this.

Comment (ejona86) [inline settings.gradle]: Oh, except we also need this flag on doing distribution binaries. It probably still makes sense for us to need the flag set in our environment to enable this part of the build, though.

**Comment by ejona86** [comment]
Looks good. I think we just need to decide what the default setting will be.

File: settings.gradle (hunk 2 above)
Comment (zhangkun83) [inline settings.gradle]: Yes. I was also thinking about continuous build/testing. I kind of lean to not skipping codegen by default.

Comment (zhangkun83) [inline settings.gradle]: Skipping the codegen by default is risky, while people would not mind adding an option to their gradle.properties, which is a lot easier than setting up C++ build environment and compiling protobuf.

Comment (ejona86) [inline settings.gradle]: Okay, fair enough. Sounds good.

**Comment by ejona86** [comment]
@zhangkun83, LGTM, but the netty submodule change needs to be removed.

**Comment by madongfly** [comment]
After this PR, "gradle clean" will delete the generated files that are tracked, which is annoy, do we have a solution for that?

**Comment by ejona86** [comment]
Yeah, it is a bit annoying. I've just done a `git checkout examples` or similar to put the codegen back in place. I don't know if there are any good solutions though.

**Pull request merged**